src/coverlet.template/ModuleTrackerTemplate.cs

```diff
@@ -38,15 +32,18 @@ static ModuleTrackerTemplate()
 
         public static void RecordHit(int hitLocationIndex)
         {
-            if (s_isTracking)
+            // Make sure to avoid recording if this is a call to RecordHit within the AppDomain setup code in an
+            // instrumented build of System.Private.CoreLib.
+            if (HitsArray is null)
                 return;
 
             Interlocked.Increment(ref HitsArray[hitLocationIndex]);
         }
 
         public static void UnloadModule(object sender, EventArgs e)
         {
-            s_isTracking = true;
+            // Claim the current hits array and reset it to prevent double-counting scenarios.
+            var hitsArray = Interlocked.Exchange(ref HitsArray, new int[HitsArray.Length]);
```

Comment:
Shouldn't HitsArray be set to null here in the exchange to avoid RecordHit doing any additional counting in CoreLib now when UnloadModule is called?

Reply:
Additional counting is safe and not inherently a bug. This approach guarantees that if the null check in `RecordHit` passes, then the access of an array element on the next line will also see a non-null array.

Reply:
True, those counts will be ignored since it's in a new array.